### ISSUE TITLE
Implementation of Self-Attention with Relative Position Representations

### ISSUE THREAD

**Issue by ksbsk** [allenai/allennlp#3398]
Hi,

Lately I've been working on an implementation of Relative Position Representations (RPR), proposed by [Shaw et al. (2018)](https://arxiv.org/pdf/1803.02155.pdf), for the Transformer model. By default the Transformer model in AllenNLP uses sinusoidal position encodings, as in the [original paper by Vaswani et al. (2017)](https://arxiv.org/pdf/1706.03762.pdf), or is not provided any position information at all:
https://github.com/allenai/allennlp/blob/1852f6130cafbef51a9cd3ec76dfe8d9dbd76f20/allennlp/modules/seq2seq_encoders/stacked_self_attention.py#L136-L140
Currently, RPR is implemented in tensor2tensor library but not in fairseq, nor it's on their roadmap ([see here](https://github.com/pytorch/fairseq/issues/556)). Which is a pity, because RPR is an interesting addition to self-attention and improves performance for NLP tasks such as machine translation, for instance. Also, RPR addresses an inherent limitation of the self-attention mechanism of the Transformer, as it considers arbitrary relations between any two elements of the input.
However, before proposing a PR, in light of recent discussions on Transformer models in AllenNLP (e.g. [[1](https://github.com/allenai/allennlp/issues/2270)], [[2](https://github.com/allenai/allennlp/issues/3334)]), I would like to confirm with you whether it would be a good idea to add RPR to [allennlp.modules.seq2seq_encoders.stacked_self_attention.StackedSelfAttentionEncoder](https://allenai.github.io/allennlp-docs/api/allennlp.modules.seq2seq_encoders.html?highlight=self%20attention#module-allennlp.modules.seq2seq_encoders.stacked_self_attention)

Let me know your thoughts!

**Comment by schmmd** [comment]
Thanks @ksbsk for the offer.  @matt-gardner would you welcome this addition to the library?

**Comment by matt-gardner** [comment]
@ksbsk, what does the default pytorch implementation of transformers do?  I think our recommendation is to use pytorch's implementation going forward.  Do they implement this option?  If they don't, I think it's probably best to submit your PR there - having a single centralized implementation of the transformer architecture is better than having us maintain a separate one. 
E.g., we don't have our own implementation of an LSTM - pytorch had a good one, there was no need for us to have a separate one.  Pytorch now has a good transformer implementation, so we shouldn't need to maintain our own.  We should probably remove all of our implementations of it for the 1.0 release.

**Comment by ksbsk** [comment]
Thanks @matt-gardner for clarifying the situation. Indeed, starting from PyTorch v1.2 release the `nn.Transformer` module is available. However, I'm not aware of any official implementation of RPR for PyTorch. What is more, it seems like `PositionalEncoding` module, covered in the original paper, was excluded from the main Transformer module at the [code peer review stage](https://github.com/pytorch/pytorch/pull/19274) and left to the user or possibly a future [PR(?)](https://github.com/pytorch/pytorch/issues/24826). At the moment, I believe PyTorch's implementation of `PositionalEncoding` module is only available through their [tutorial](https://pytorch.org/tutorials/beginner/transformer_tutorial.html?highlight=transformer).

So the bottom line is that: so far no signs of the standard RPR module implementation.

**Comment by nelson-liu** [comment]
For the reference, there are no signs of a standard positional embedding module implementation in the PyTorch library itself, right? Positional embeddings (either absolute or relative) seem pretty crucial for performance (and they're included in the original paper as well).

My read on why it wasn't included natively is because the positional embeddings are technically an external component of the model--namely, it feels like a `TextFieldEmbedder` (but this can be debated. For instance, the AllenNLP transformer itself includes positional embeddings https://github.com/allenai/allennlp/blob/3dda5ac9fcf390d8b83eb855249360711707c11c/allennlp/modules/seq2seq_encoders/bidirectional_language_model_transformer.py#L48_L67 ).

I think it'd be useful to have some sort of `TextFieldEmbedder` that takes in an input text and generates positional embeddings, so it's easy for someone to, say, use the pytorch transformer model with concatenated configurable positional embeddings. Or, we should make a wrapper of the transformer model that adds positional embeddings (since I'd guess that the majority of people who want to use the transformer want positional embeddings. This was the assumption in our current Transformer code, but this is not true of the PyTorch transformer code).

tl;dr: pytorch transformer code does not position embeddings, while allennlp transformer does. Position embeddings seem important, and I'd be strongly in favor of having them in the library.

**Comment by matt-gardner** [comment]
Thanks for the additional context @nelson-liu!  I think I agree with your assessment, also, that this is likely best handled as an additional `TokenEmbedder` (I believe that's what you meant, instead of `TextFieldEmbedder`).  We could have a `TokenIndexer` that outputs an integer token position index, and a `TokenEmbedder` that takes those indices and outputs a position encoding (and the `TokenEmbedder` would just be a module that maps integers to embeddings that could also be used in other contexts).

@ksbsk, if you agree with this framing of it, PR is welcome!

**Comment by nelson-liu** [comment]
Oops yes I meant TokenEmbedder . Come to think of it, it actually just might make sense as an optional constructor argument of a PytorchTransformerWrapper Seq2Seq / Seq2Vec encoder...

**Comment by matt-gardner** [comment]
I wouldn't bake it into those abstractions.  It would be easy enough to instantiate the module yourself inside your model and use it in between layers wherever you want.

**Comment by ruleGreen** [comment]
so how is it going now?

**Comment by DeNeutoy** [comment]
We should probably defer to transformers on this point, as we are trying to reduce the amount of transformer implementations we have inside allennlp currently: #2270

**Comment by Wufang1997** [comment]
You can refer to https://github.com/TensorUI/relative-position-pytorch/blob/master/relative_position.py. It's a good practice of RPE.
